x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinTypes.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.esql.plan.logical.join;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xpack.esql.core.capabilities.Resolvables;
+import org.elasticsearch.xpack.esql.core.expression.Attribute;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Utility class defining the concrete types of joins supported by ESQL.
+ */
+public class JoinTypes {
+
+    private JoinTypes() {}
+
+    public static JoinType INNER = CoreJoinType.INNER;
+    public static JoinType LEFT = CoreJoinType.LEFT;
+    public static JoinType RIGHT = CoreJoinType.RIGHT;
+    public static JoinType FULL = CoreJoinType.FULL;
+    public static JoinType CROSS = CoreJoinType.CROSS;
+
+    /**
+     * The predefined core join types. Implements as enum for easy comparison and serialization.
+     */
+    private enum CoreJoinType implements JoinType {
```

Comment:
This class implementing `JoinType` feels a bit odd -- does it need to? 
I _suppose_ it's so to be able to keep it private, so that the public constants can be used as is? But it feels a bit odd to init other implementations of `JoinType` (like  `UsingJoinType`) with a `JoinType`, to actually defines its behaviour.
In the Analyser we check for an instance of `UsingJoinType`, which is a `JoinType`, then get out of it the core type, which is another `JoinType`.
I guess what I'm wondering is if we could model the concepts of "how do we stitch" vs "what do we match" separately?

Reply:
It's unusual - trying to mix the enum/limited functionality with extension for holding the keys specified by USING.
LEFT JOIN x USING a is different from LEFT JOIN x ON l.a == x.b (assuming namespace will be supported) - it's not just about the condition but also semantics (SQL USING returns the keys as the first column).

There's no strong reason to keep the core enum private; mainly to force the use of the interface and JoinTypes constants helps with evolution (before the enum was differently defined).